specs/kyber/src/sampling.rs

```diff
@@ -0,0 +1,141 @@
+///     end if
+///     i := i + 3
+/// end while
+/// return aˆ0 + aˆ1X + · · · + aˆn−1Xn−1
+/// ```
+///
+/// The Kyber Round 3 specification can be found at:
+/// https://pq-crystals.org/kyber/data/kyber-specification-round3-20210131.pdf
+///
+pub(crate) fn sample_ring_element_uniform(randomness : [u8; parameters::REJECTION_SAMPLING_SEED_SIZE]) -> Result<RingElement, &'static str> {
+    let mut j : usize = 0;
+    let mut sampled : RingElement = RingElement::ZERO;
+
+    for i in (0..randomness.len()).step_by(3) {
+        let byte = u16::from(randomness[i]);
+        let byte1 = u16::from(randomness[i + 1]);
+        let byte2 = u16::from(randomness[i + 2]);
+
+        let d1 = byte + (256 * (byte1 % 16));
+
+        // Integer division is flooring in Rust.
+        let d2 = (byte1 / 16) + (16 * byte2); 
+
+        if d1 < parameters::FIELD_MODULUS && j < sampled.coefficients.len() {
+            sampled.coefficients[j] = FieldElement::from_u16(d1);
+            j += 1
+        }
+        if d2 < parameters::FIELD_MODULUS && j < sampled.coefficients.len() {
+            sampled.coefficients[j] = FieldElement::from_u16(d2);
+            j += 1;
+        }
+        // In an efficient implementation, we'd break when
+        // j == sampled.coefficients.len, but we've forgone this to make
+        // the implementation easier to read.
+    }
+
+    if j == sampled.coefficients.len() {
+        Ok(sampled)
+    } else {
+        Err("Not enough randomness to sample the ring element.")
+    }
+}
+
+///
+/// Given a series of uniformly random bytes in |randomness|, sample
+/// a ring element from a binomial distribution centered at 0 that uses two sets
+/// of |parameters::BINOMIAL_SAMPLING_COINS| coin flips.
+///
+/// This function implements Algorithm 2 of the Kyber Round 3 specification,
+/// which is reproduced below:
+///
+/// ```plaintext
+/// Input: Byte array B = (b0, b1, . . . , b64η−1) ∈ B64η
+/// Output: Polynomial f ∈ Rq
+/// (β0, . . . , β512η−1) := BytesToBits(B)
+/// for i from 0 to 255 do
+///     a := sum(j = 0 to η−1)(β2iη+j)
+///     b := sum(j = 0 to η−1)(β2iη+η+j)
+///     fi := a − b
+/// end for
+/// return f0 +f1X +f2X2 +···+f255X255
+/// ```
+///
+/// The Kyber Round 3 specification can be found at:
+/// https://pq-crystals.org/kyber/data/kyber-specification-round3-20210131.pdf
+///
+pub(crate) fn sample_ring_element_binomial(randomness : [u8; parameters::BINOMIAL_SAMPLING_COINS * 64]) -> RingElement {
+    let random_bits = bytes_to_bit_vector(&randomness[..]);
+    assert_eq!(random_bits.len(), parameters::BINOMIAL_SAMPLING_COINS * 64 * (u8::BITS as usize));
+
+    let mut sampled : RingElement = RingElement::ZERO;
+
+    for i in 0..sampled.coefficients.len() {
+        let mut a : u8 = 0;
+        for j in 0..parameters::BINOMIAL_SAMPLING_COINS {
+            a += random_bits[(2 * i) * parameters::BINOMIAL_SAMPLING_COINS + j];
+        }
+        let coin_tosses_a = FieldElement::from_u16(u16::from(a));
```

Comment:
A `from_u8` would be nicer here.
In general you could as well implement the `Into` trait to make the spec a little nicer for all these convertions.